File: packages/protocol/src/logs.ts
/**
 * Application log records carried over the semantic channel.
 *
 * A TUI cannot print diagnostics to the screen without corrupting the render,
 * so applications write them to an internal logger instead. The `logs`
 * capability lets an instrumented adapter forward those records to the driver,
 * where they become assertable test state rather than invisible side effects.
 *
 * Records are bounded exactly like snapshots: projected into frozen plain DTOs
 * before retention, checked against a byte ceiling, and rejected wholesale on
 * any violation. A misbehaving logger degrades into dropped records, never
 * into unbounded driver memory.
 */

import { Buffer } from 'node:buffer';
import type { ProtocolLimits } from './limits.js';
import type { ValidationErrorCode } from './validate.js';
import { ProtocolViolation } from './errors.js';
import { projectDto } from './framing.js';

/**
 * Severity ladder, ordered from least to most severe. Deliberately the
 * intersection of the ladders used by pino, winston, consola, Python
 * `logging`, Go `slog` and Rust `tracing`, so every bridge maps onto it
 * without inventing a level.
 */
export const LOG_LEVELS = ['trace', 'debug', 'info', 'warn', 'error', 'fatal'] as const;

export type LogLevel = (typeof LOG_LEVELS)[number];

/** Numeric severity, useful for threshold comparisons. Higher is more severe. */
export const LOG_LEVEL_SEVERITY: Readonly<Record<LogLevel, number>> = Object.freeze({
  trace: 10,
  debug: 20,
  info: 30,
  warn: 40,
  error: 50,
  fatal: 60,
});

/**
 * Structured attribute value. Scalars only, by design: nested objects make
 * record size unbounded and depth-dependent, and every bridge already has to
 * flatten for its own transport. `@termwright/logs` does the flattening.
 */
export type LogAttrValue = string | number | boolean | null;

/** Maximum number of attribute keys on one record. */
export const MAX_LOG_ATTRS = 64;

/**
 * One application log record.
 *
 * @remarks
 * `ts` is **Unix epoch milliseconds**, not session-relative: the adapter has no
 * reliable view of when the driver considers the session to have started, so
 * the only clock both sides can agree on without negotiation is the wall
 * clock. The driver rebases it onto the session/cast timeline.
 */
export interface LogRecord {
  /** Unix epoch milliseconds when the record was produced. */
  readonly ts: number;
  readonly level: LogLevel;
  /** Human-readable message, already formatted by the source logger. */
  readonly message: string;
  /** Flat structured context. Nested values are flattened by the bridge. */
  readonly attrs?: Readonly<Record<string, LogAttrValue>>;
  /** Logger/channel name, e.g. `http` or `db.pool`. */
  readonly logger?: string;
  /**
   * Per-session counter, non-decreasing, assigned by the adapter. A gap tells
   * the driver records were dropped upstream (rate limit, queue overflow)
   * rather than lost in transit.
   */
  readonly seq: number;
  /** Semantic revision current when the record was produced, when known. */
  readonly revision?: number;
}

/** Structured result: never throws hostile data onward. */
export type LogValidationResult =
  | { readonly ok: true; readonly record: LogRecord }
  | { readonly ok: false; readonly code: ValidationErrorCode; readonly detail: string };

function fail(code: ValidationErrorCode, detail: string): LogValidationResult {
  return { ok: false, code, detail };
}

const LEVELS: ReadonlySet<string> = new Set(LOG_LEVELS);

function isSafeNonNegative(value: unknown): value is number {
  return typeof value === 'number' && Number.isSafeInteger(value) && value >= 0;
}

/**
 * Validate an untrusted log record.
 *
 * Mirrors {@link import('./validate.js').validateSnapshot}: the value is
 * projected into a frozen plain DTO first (so getters are rejected without
 * being invoked), then measured against the byte ceiling, then checked field
 * by field.
 *
 * @param value - Untrusted candidate record.
 * @param limits - Active limits; `maxLogRecordBytes` and `maxStringBytes` apply.
 * @returns `{ ok: true, record }` with a deep-frozen record, or a typed
 * failure. Never throws.
 */
export function validateLogRecord(value: unknown, limits: ProtocolLimits): LogValidationResult {
  let projected: unknown;
  try {
    projected = projectDto<unknown>(value, limits.maxDepth);
  } catch (error) {
    if (error instanceof ProtocolViolation) {
      return fail(error.code === 'dto-depth' ? 'depth' : 'schema', error.message);
    }
    return fail('schema', 'value could not be projected into a plain DTO');
  }

  const serialised = JSON.stringify(projected);
  if (serialised === undefined) {
    return fail('schema', 'log record is not a JSON object');
  }
  const bytes = Buffer.byteLength(serialised, 'utf8');
  if (bytes > limits.maxLogRecordBytes) {
    return fail('bytes', `log record is ${bytes} bytes, ceiling is ${limits.maxLogRecordBytes}`);
  }

  if (typeof projected !== 'object' || projected === null || Array.isArray(projected)) {
    return fail('schema', 'log record must be an object');
  }
  const record = projected as Record<string, unknown>;

  for (const key of Object.keys(record)) {
    if (!['ts', 'level', 'message', 'attrs', 'logger', 'seq', 'revision'].includes(key)) {
      return fail('schema', `unknown log record property "${key}"`);
    }
  }

  if (!isSafeNonNegative(record['ts']) || record['ts'] === 0) {
    return fail('schema', 'ts must be a positive safe integer (epoch milliseconds)');
  }
  if (typeof record['level'] !== 'string' || !LEVELS.has(record['level'])) {
    return fail('schema', `level must be one of ${LOG_LEVELS.join(', ')}`);
  }
  if (typeof record['message'] !== 'string') {
    return fail('schema', 'message must be a string');
  }
  if (Buffer.byteLength(record['message'], 'utf8') > limits.maxStringBytes) {
    return fail('string-bytes', `message exceeds ${limits.maxStringBytes} UTF-8 bytes`);
  }
  if (!isSafeNonNegative(record['seq'])) {
    return fail('schema', 'seq must be a non-negative safe integer');
  }

  if (record['logger'] !== undefined) {
    if (typeof record['logger'] !== 'string') {
      return fail('schema', 'logger must be a string');
    }
    if (Buffer.byteLength(record['logger'], 'utf8') > limits.maxStringBytes) {
      return fail('string-bytes', `logger exceeds ${limits.maxStringBytes} UTF-8 bytes`);
    }
  }

  if (record['revision'] !== undefined) {
    if (!isSafeNonNegative(record['revision']) || record['revision'] === 0) {
      return fail('revision', 'revision must be a positive safe integer');
    }
  }

  const attrs = record['attrs'];
  if (attrs !== undefined) {
    if (typeof attrs !== 'object' || attrs === null || Array.isArray(attrs)) {
      return fail('schema', 'attrs must be a flat object');
    }
    const entries = Object.entries(attrs as Record<string, unknown>);
    if (entries.length > MAX_LOG_ATTRS) {
      return fail('count', `attrs carries ${entries.length} keys, ceiling is ${MAX_LOG_ATTRS}`);
    }
    for (const [key, attrValue] of entries) {
      if (Buffer.byteLength(key, 'utf8') > limits.maxStringBytes) {
        return fail('string-bytes', `attribute key "${key}" exceeds the string ceiling`);
      }
      const type = typeof attrValue;
      if (attrValue !== null && type !== 'string' && type !== 'number' && type !== 'boolean') {
        return fail('schema', `attribute "${key}" must be a string, number, boolean or null`);
      }
      if (type === 'number' && !Number.isFinite(attrValue)) {
        return fail('schema', `attribute "${key}" must be a finite number`);
      }
      if (type === 'string' && Buffer.byteLength(attrValue as string, 'utf8') > limits.maxStringBytes) {
        return fail('string-bytes', `attribute "${key}" exceeds the string ceiling`);
      }
    }
  }

  return { ok: true, record: projected as LogRecord };
}

File: clients/python/src/termwright/logs.py
"""Application log records carried over the semantic channel.

A TUI cannot print diagnostics to the screen without corrupting the render, so
applications write them to a logger instead. The ``logs`` capability forwards
those records to the driver, where they become assertable test state rather
than invisible side effects.

Records are bounded exactly like snapshots: checked against a byte ceiling and
rejected wholesale on any violation, so a misbehaving logger degrades into
dropped records rather than unbounded driver memory.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Dict, Mapping, Optional, Union

from .framing import encode_json, project_dto
from .errors import ProtocolViolation
from .limits import DEFAULT_LIMITS, ProtocolLimits

#: Severity ladder, least to most severe. The intersection of the ladders used
#: by Python ``logging``, Go ``slog``, Rust ``tracing``, pino and winston, so
#: every bridge maps onto it without inventing a level.
LOG_LEVELS = ("trace", "debug", "info", "warn", "error", "fatal")

LEVEL_SET = frozenset(LOG_LEVELS)

#: Numeric severity; higher is more severe.
LOG_LEVEL_SEVERITY: Dict[str, int] = {
    "trace": 10,
    "debug": 20,
    "info": 30,
    "warn": 40,
    "error": 50,
    "fatal": 60,
}

#: Maximum number of attribute keys on one record.
MAX_LOG_ATTRS = 64

RECORD_FIELDS = ("ts", "level", "message", "attrs", "logger", "seq", "revision")

_MAX_SAFE_INTEGER = 2**53 - 1

#: Values an attribute may hold. Scalars only: nested objects make record size
#: unbounded and depth-dependent, and every bridge already flattens for its own
#: transport.
LogAttrValue = Union[str, int, float, bool, None]


@dataclass(frozen=True)
class LogRecord:
    """One application log record.

    ``ts`` is Unix epoch milliseconds, not session-relative: an adapter has no
    reliable view of when the driver considers the session to have started, so
    the wall clock is the only clock both sides agree on without negotiating.
    The driver rebases it onto the session timeline.
    """

    ts: int
    level: str
    message: str
    seq: int
    attrs: Optional[Mapping[str, LogAttrValue]] = None
    logger: Optional[str] = None
    revision: Optional[int] = None

    def to_wire(self) -> Dict[str, Any]:
        """Serialise, dropping unset optionals: the schema is strict."""
        wire: Dict[str, Any] = {"ts": self.ts, "level": self.level, "message": self.message}
        if self.attrs is not None:
            wire["attrs"] = dict(self.attrs)
        if self.logger is not None:
            wire["logger"] = self.logger
        wire["seq"] = self.seq
        if self.revision is not None:
            wire["revision"] = self.revision
        return wire


@dataclass(frozen=True)
class LogValidationResult:
    """Outcome of :func:`validate_log_record`."""

    ok: bool
    record: Optional[Dict[str, Any]] = None
    code: Optional[str] = None
    detail: str = ""


def _fail(code: str, detail: str) -> LogValidationResult:
    return LogValidationResult(ok=False, code=code, detail=detail)


def _utf8_len(text: str) -> int:
    return len(text.encode("utf-8", "surrogatepass"))


def _safe_non_negative(value: Any) -> bool:
    return (
        isinstance(value, int)
        and not isinstance(value, bool)
        and 0 <= value <= _MAX_SAFE_INTEGER
    )


def validate_log_record(
    value: Any, limits: ProtocolLimits = DEFAULT_LIMITS
) -> LogValidationResult:
    """Validate an untrusted log record against ``limits``.

    Mirrors :func:`termwright.validate.validate_snapshot`: the value is
    projected first, then measured against ``maxLogRecordBytes``, then checked
    field by field. Never raises.
    """
    try:
        projected = project_dto(value, limits.maxDepth)
    except ProtocolViolation as error:
        return _fail("depth" if error.code == "dto-depth" else "schema", str(error))

    try:
        serialised = encode_json(projected)
    except ProtocolViolation:
        return _fail("schema", "log record is not JSON-serialisable")
    if len(serialised) > limits.maxLogRecordBytes:
        return _fail(
            "bytes",
            f"log record is {len(serialised)} bytes, ceiling is {limits.maxLogRecordBytes}",
        )

    if not isinstance(projected, dict):
        return _fail("schema", "log record must be an object")
    record: Dict[str, Any] = projected

    for key in record:
        if key not in RECORD_FIELDS:
            return _fail("schema", f'unknown log record property "{key}"')

    if not _safe_non_negative(record.get("ts")) or record["ts"] == 0:
        return _fail("schema", "ts must be a positive safe integer (epoch milliseconds)")
    if record.get("level") not in LEVEL_SET:
        return _fail("schema", f"level must be one of {', '.join(LOG_LEVELS)}")
    if not isinstance(record.get("message"), str):
        return _fail("schema", "message must be a string")
    if _utf8_len(record["message"]) > limits.maxStringBytes:
        return _fail("string-bytes", f"message exceeds {limits.maxStringBytes} UTF-8 bytes")
    if not _safe_non_negative(record.get("seq")):
        return _fail("schema", "seq must be a non-negative safe integer")

    if "logger" in record:
        if not isinstance(record["logger"], str):
            return _fail("schema", "logger must be a string")
        if _utf8_len(record["logger"]) > limits.maxStringBytes:
            return _fail("string-bytes", f"logger exceeds {limits.maxStringBytes} UTF-8 bytes")

    if "revision" in record:
        if not _safe_non_negative(record["revision"]) or record["revision"] == 0:
            return _fail("revision", "revision must be a positive safe integer")

    if "attrs" in record:
        attrs = record["attrs"]
        if not isinstance(attrs, dict):
            return _fail("schema", "attrs must be a flat object")
        if len(attrs) > MAX_LOG_ATTRS:
            return _fail("count", f"attrs carries {len(attrs)} keys, ceiling is {MAX_LOG_ATTRS}")
        for key, attr in attrs.items():
            if _utf8_len(key) > limits.maxStringBytes:
                return _fail("string-bytes", f'attribute key "{key}" exceeds the string ceiling')
            if attr is not None and not isinstance(attr, (str, int, float, bool)):
                return _fail(
                    "schema", f'attribute "{key}" must be a string, number, boolean or null'
                )
            if isinstance(attr, float) and (attr != attr or attr in (float("inf"), float("-inf"))):
                return _fail("schema", f'attribute "{key}" must be a finite number')
            if isinstance(attr, str) and _utf8_len(attr) > limits.maxStringBytes:
                return _fail("string-bytes", f'attribute "{key}" exceeds the string ceiling')

    return LogValidationResult(ok=True, record=record)


def flatten_attrs(
    value: Mapping[str, Any], prefix: str = "", depth: int = 0
) -> Dict[str, LogAttrValue]:
    """Flatten nested context into dotted keys, as the wire format requires.

    ``{"db": {"host": "x"}}`` becomes ``{"db.host": "x"}``. Values that are not
    scalars after flattening are rendered with :func:`repr`, because losing the
    shape of a value is better than dropping the record that carries it.
    """
    flat: Dict[str, LogAttrValue] = {}
    for key, item in value.items():
        name = f"{prefix}{key}"
        if isinstance(item, Mapping) and depth < 4:
            flat.update(flatten_attrs(item, f"{name}.", depth + 1))
        elif item is None or isinstance(item, (str, bool, int)):
            flat[name] = item
        elif isinstance(item, float):
            flat[name] = item if item == item and item not in (float("inf"), float("-inf")) else str(item)
        else:
            try:
                flat[name] = json.dumps(item, default=repr)
            except (TypeError, ValueError):
                flat[name] = repr(item)
    return flat
